### PR TITLE
net/raft: add unit tests for node eviction

### PR DESCRIPTION
**Pull request by vickiniu**
Adds two unit tests for node eviction: for evicting a single node, and multiple nodes. Also slightly refactors to add a helper function to create a test Raft cluster.

### CODE REVIEW

File: net/raft/raft_test.go [hunk 1]
```diff
@@ -157,6 +136,53 @@ func TestClusterSetup(t *testing.T) {
 	if got != "bar" {
 		t.Errorf("reading /foo, nodeC got %q want %q", got, "bar")
 	}
+	nodeA.cleanup()
```
Comment (kr) [inline net/raft/raft_test.go]: It's probably better to `defer` each of these calls just after line 127 where the nodes are created.

File: net/raft/raft_test.go [hunk 2]
```diff
@@ -157,6 +136,53 @@ func TestClusterSetup(t *testing.T) {
...
+
+func TestEvictMultiple(t *testing.T) {
+	ctx := context.Background()
+
+	// Create new test cluster
+	nodeA, nodeB, nodeC := newTestCluster(ctx, t)
+	addrB := nodeB.addr
+	addrC := nodeC.addr
+
+	// Have nodeC evict nodeA and nodeB
```
Comment (kr) [inline net/raft/raft_test.go]: This comment doesn't seem to match what the test is doing (nodeA is evicting the other two).

Comment (kr) [inline net/raft/raft_test.go]: But along the lines of this comment, would it be worth testing that a node other than node A can evict the other two (either here on in an additional test)?

Comment (vickiniu) [inline net/raft/raft_test.go]: If we try to evict node A, this triggers a panic from raft.go:738

```go
if cc.NodeID == sv.id {
	panic(errors.New("removed from cluster"))
}
```

I'm not sure what this piece of code is meant to do, but removing this check means that evicting the leader works. Do we want to keep this check in?

**Comment by vickiniu** [comment]
etcd example for dropping messages w/o matching peer addresses: https://github.com/coreos/etcd/blob/e962b0c8498719c8d14c06572c38907ef4919343/etcdserver/raft.go#L302

File: net/raft/http.go [hunk 3]
```diff
@@ -161,6 +168,17 @@ func sendmsg(addr string, data []byte, client *http.Client) {
...
+// Sends eviction notice to a node to kill the process
+func evictionNotice(addr string, client *http.Client) {
+	url := "https://" + addr + "/raft/kill"
+	resp, err := client.Post(url, contentType, nil)
+	if err != nil {
+		log.Printkv(context.Background(), "warning", err)
+		return
+	}
+	defer resp.Body.Close()
+}
```
Comment (jbowens) [inline net/raft/http.go]: I wonder if instead of pushing eviction notices to evicted nodes, we should provide facilities for a node to recognize that it's evicted when it performs a `sendmsg` to contact other nodes?

Since these eviction notices are best-effort, it's still possible for an eviction to go unnoticed. For example:
1. Node _A_ is the Chain Core leader process.
2. A network partition causes node _A_ to be unreachable from the other processes.
3. An operator intervenes and evicts node _A_ from the cluster.
4. All the other nodes process the eviction entry and try to deliver the eviction notice to node _A_. Because node _A_ is unreachable, they all fail.
5. The network is restored.
6. Node _A_ is still evicted, but its Chain Core process is still running.

Node _A_ would continue to be Chain Core leader but without a working raft cluster.

I guess this scenario could also happen if the network partition is never restored and the node is never evicted. The leader process could continue despite not being able to participate in raft. Maybe the real solution is just to move Chain Core leadership into raft.

Comment (vickiniu) [inline net/raft/http.go]: Just updated to have nodes recognize evictions when they get a response that they're no longer in the cluster — thanks!

File: net/raft/http.go [hunk 4]
```diff
@@ -161,6 +185,17 @@ func sendmsg(addr string, data []byte, client *http.Client) {
 	defer resp.Body.Close()
 }
 
+// Sends eviction notice to a node to kill the process
+func evictionNotice(addr string, client *http.Client) {
```
Comment (tessr) [inline net/raft/http.go]: I think you can remove this function now, right?

File: net/raft/raft.go [hunk 5]
```diff
@@ -735,10 +745,10 @@ func (sv *Service) applyEntry(ent raftpb.Entry, writers map[string]chan bool) {
 		case raftpb.ConfChangeAddNode, raftpb.ConfChangeUpdateNode:
 			sv.state.SetPeerAddr(cc.NodeID, string(cc.Context))
 		case raftpb.ConfChangeRemoveNode:
+			sv.state.RemovePeerAddr(cc.NodeID)
 			if cc.NodeID == sv.id {
-				panic(errors.New("removed from cluster"))
+				os.Exit(0)
```
Comment (tessr) [inline net/raft/raft.go]: why this change?

Comment (vickiniu) [inline net/raft/raft.go]: This was basically for testing: `panic` would cause the test to crash with an error and I couldn't `recover` from the test code since the `panic` was in a different goroutine. Also, since this is expected behavior when a node is removed from the cluster, I'm not sure if a panic makes more sense than just exiting.

File: net/raft/http.go (hunk 4 above)
Comment (vickiniu) [inline net/raft/http.go]: Yes, definitely! Thanks for the catch

**Comment by vickiniu** [comment]
PTAL :)

File: net/raft/raft.go [hunk 6]
```diff
@@ -407,6 +407,15 @@ func (sv *Service) runUpdatesReady(rd raft.Ready, wal *wal.WAL, writers map[stri
 	sv.raftNode.Advance()
 }
 
+func (sv *Service) processMessages(msgs []raftpb.Message) []raftpb.Message {
```
Comment (tessr) [inline net/raft/raft.go]: could you add a comment here explaining what is going on, please?

Comment (vickiniu) [inline net/raft/raft.go]: added!

**Comment by tessr** [comment]
LGTM
